components/sup/src/manager/service/pipe_hook_client.rs

```diff
@@ -0,0 +1,319 @@
+use habitat_common::{error::{Error,
+                             Result},
+                     outputln,
+                     templating::package::Pkg};
+use habitat_core::{env as henv,
+                   os::process::windows_child::Child};
+use mio::{Events,
+          Poll,
+          PollOpt,
+          Ready,
+          Token};
+use mio_named_pipes::NamedPipe;
+use std::{self,
+          env,
+          ffi::OsStr,
+          fs::{File,
+               OpenOptions},
+          io::{self,
+               prelude::*,
+               BufRead,
+               BufReader,
+               Read},
+          iter::once,
+          os::windows::{ffi::OsStrExt,
+                        fs::*,
+                        io::*},
+          path::PathBuf,
+          thread,
+          time::{Duration,
+                 Instant}};
+use uuid::Uuid;
+use winapi::um::{namedpipeapi,
+                 processthreadsapi,
+                 winbase};
+
+static LOGKEY: &str = "HK";
+const PIPE_SERVICE_ENVVAR: &str = "HAB_HOOK_PIPE_SCRIPT";
+const START_PIPE_TIMEOUT_SECONDS: u64 = 5;
+const SIGNAL_EXEC_HOOK: [u8; 1] = [1];
+const SIGNAL_QUIT: [u8; 1] = [0];
+
+#[derive(Debug)]
+pub struct PipeHookClient {
+    hook_name:       String,
+    hook_path:       PathBuf,
+    pipe_name:       String,
+    stdout_log_file: PathBuf,
+    stderr_log_file: PathBuf,
+}
+
+impl PipeHookClient {
+    pub fn new(hook_name: String,
+               hook_path: PathBuf,
+               stdout_log_file: PathBuf,
+               stderr_log_file: PathBuf)
+               -> Self {
+        let pipe_name = format!("{}-hook-{}", hook_name, Uuid::new_v4());
+        Self { hook_name,
+               hook_path,
+               pipe_name,
+               stdout_log_file,
+               stderr_log_file }
+    }
+
+    pub fn exec_hook<T>(&self,
+                        service_group: &str,
+                        pkg: &Pkg,
+                        svc_encrypted_password: Option<T>)
+                        -> Result<u32>
+        where T: ToString
+    {
+        // Always check to see if the powershell named pipe server is running.
+        // This should only be false the first time this function is called
+        // or if someone explicitly terminates the powershell process
+        if self.pipe_exists().is_err() {
```

Comment:
Well we expect it to exist on the 2nd call to `exec_hook` and onwards. However, its vwery possible someone can inadvertantly terminate the hosting `pwsh` process and obliterate the pipe. If there is something fundamentally wrong and we can't create the pipe, the call to `init_server` should err resulting in killing the supervisor. If for some reason the client did was not propperly disconnected from the pipe on the last call to `exec_hook`, the call to `init_server` would fail trying to create a new pipe that is busy. However we have 2 protections against this scenario:
1. the client's pipe handle is dropped and thus closed at the end of `exec_hook`
2. the powershell server disconnects the client after returning its exit code.